qst/src/main/java/io/deephaven/qst/table/NewTable.java

```diff
@@ -0,0 +1,150 @@
+package io.deephaven.qst.table;
+
+import io.deephaven.qst.LeafStyle;
+import io.deephaven.qst.array.Array;
+import io.deephaven.qst.column.Column;
+import io.deephaven.qst.column.header.ColumnHeader;
+import org.immutables.value.Value.Check;
+import org.immutables.value.Value.Immutable;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.Spliterator;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+@Immutable
+@LeafStyle
+public abstract class NewTable extends TableBase implements Iterable<Column<?>> {
+    public interface Builder {
+
+        Builder size(int size);
+
+        Builder putColumns(String name, Array<?> value);
+
+        default Builder addColumns(Column<?> column) {
+            putColumns(column.name(), column.array());
+            return this;
+        }
+
+        default Builder addColumns(Column<?>... columns) {
+            for (Column<?> column : columns) {
+                addColumns(column);
+            }
+            return this;
+        }
+
+        default Builder addAllColumns(Iterable<Column<?>> columns) {
+            for (Column<?> column : columns) {
+                addColumns(column);
+            }
+            return this;
+        }
+
+        NewTable build();
+    }
+
+    public static Builder builder() {
+        return ImmutableNewTable.builder();
+    }
+
+    public static NewTable empty(TableHeader header) {
```

Comment:
We're overloading "empty". I know it's weird, but `EmptyTable` is column-less. Maybe we should just allow `NewTable` with no columns...

Reply:
Defer for now, ?

Reply:
Ticket created, and referenced in another related comment.